### PR TITLE
Do some foreign key tings

### PR DESCRIPTION
**Pull request by xd009642**

### CODE REVIEW

File: tests/student_database.rs [hunk 1]
```diff
@@ -65,9 +73,27 @@ impl StudentDb {
 #[traced_test]
 fn student_simple() {
     let mut db = StudentDb::new();
+    db.init_default().unwrap();
     db.add_student("Daniel McKenna", 31, 1).unwrap();
     // Read `Building Query Compilers` it's so useful!
```
Comment (devanbenz) [inline tests/student_database.rs:78]: sounds good will do

File: src/storage_engine.rs [hunk 2]
```diff
@@ -68,7 +68,28 @@ impl StorageEngine {
         &mut self.db
     }
 
+    fn validate_table_options(&self, create_table: &CreateTableOptions) -> anyhow::Result<()> {
+        for (column, props) in create_table
+            .columns
+            .iter()
+            .filter(|(_, x)| x.foreign_key.is_some())
```
Comment (devanbenz) [inline src/storage_engine.rs:75]: I'm wondering if you could use a `filter_map` here instead to make this a little more concise:
```
.filter_map(|(_, x)| x.foreign_key.as_ref().map(|fk| (fk, x)))
```
And then you can move some of the nested if let some logic up in to the for loop call

File: src/types.rs [hunk 3]
```diff
@@ -197,8 +205,20 @@ impl TryFrom<&Statement> for Command {
...
                                 entry.unique = true;
                             }
-                            ColumnOption::ForeignKey { .. } => {
-                                anyhow::bail!("FOREIGN KEY not yet supported")
+                            ColumnOption::ForeignKey {
+                                foreign_table,
+                                referred_columns,
+                                ..
+                            } => {
+                                if referred_columns.len() != 1 {
```
Comment (devanbenz) [inline src/types.rs:213]: Should you check to see if this gets called with a 0 len value too?

File: src/types.rs [hunk 4]
```diff
@@ -197,8 +205,20 @@ impl TryFrom<&Statement> for Command {
...
+                                    anyhow::bail!(
+                                        "Exactly one column must be specified for a foreign key"
+                                    );
+                                }
+                                entry.foreign_key = Some((
+                                    foreign_table.to_string(),
+                                    referred_columns[0].to_string(),
```
Comment (devanbenz) [inline src/types.rs:220]: For these index accessors would it be better to do something like:
```
referred_columns.get(0).unwrap_or_else(....);
```

to make it a bit more safe then doing a direct array access?

File: src/types.rs [hunk 5]
```diff
@@ -215,6 +235,55 @@ impl TryFrom<&Statement> for Command {
...
+                            let name = columns[0].to_string();
+                            if let Some(column_def) = descriptor.get_mut(&name) {
+                                if referred_columns.len() != 1 {
+                                    anyhow::bail!(
+                                        "Exactly one column must be specified for a foreign key"
+                                    );
+                                }
+                                column_def.foreign_key = Some((
+                                    foreign_table.to_string(),
+                                    referred_columns[0].to_string(),
```
Comment (devanbenz) [inline src/types.rs:262]: Same here, see above.

**Review by devanbenz — CHANGES_REQUESTED** [review]

File: src/storage_engine.rs (hunk 2 above)
Comment (xd009642) [reply src/storage_engine.rs:75]: hmm I think this is kinda just moving the nesting to an extra set of method chains in a method which not sure if I prefer readability wise

File: src/types.rs (hunk 4 above)
Comment (xd009642) [reply src/types.rs:220]: There's an error check up above, so just gonna skip extra bounds checking

**Pull request merged**